File: Problem_12/AC_solution.py
n, l = map(int, input().split())
fights = list(map(int, input().split()))

min_bounty, max_bounty, ans = 0, l, 0

while min_bounty <= max_bounty:
    current_bounty = (min_bounty + max_bounty) // 2
    current_sum = 0

    for i in range(n):
        current_sum += 100 if current_bounty + current_sum > fights[i] else 0
        if current_bounty + current_sum > l:
            break

    if current_bounty + current_sum > l:
        max_bounty = current_bounty - 1
    else:
        ans = current_bounty
        min_bounty = current_bounty + 1

print(ans)

File: Problem_12/AC_solution.cpp
#include<bits/stdc++.h>  
using namespace std;
#define IOS ios::sync_with_stdio(0); cin.tie(0); cout.tie(0);

int main() {
    IOS
    int n;
    long long l;
    cin >> n >> l;
    vector<int> fights(n);
    for(int i = 0; i < n; i++) cin >> fights[i];
    long long min_bounty = 0, max_bounty = l, ans;
    while(min_bounty <= max_bounty){
        long long current_bounty = (min_bounty + max_bounty) / 2;
        long long current_sum = 0;
        for(int i = 0; i < n; i++){
            current_sum += (current_bounty + current_sum > fights[i] ? 100 : 0);
            if (current_bounty + current_sum > l) break;
        }
        if (current_bounty + current_sum > l) max_bounty = current_bounty - 1;
        else{
            ans = current_bounty;
            min_bounty = current_bounty + 1;
        }
    }
    cout << ans;
    return 0;
}